### PR TITLE
Fix newest boost and GCC warnings

### PR DESCRIPTION
**Pull request by bvbfan**
Signed-off-by: Anthony Fieroni <bvbfan@abv.bg>

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2198,12 +2198,12 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
                     else
                     {
-                        if (nChangePosInOut == -1)
+                        if (nChangePosInOut < 0)
                         {
                             // Insert change txn at random position:
                             nChangePosInOut = GetRandInt(txNew.vout.size()+1);
                         }
-                        else if (nChangePosInOut > txNew.vout.size())
+                        else if (size_t(nChangePosInOut) > txNew.vout.size())
```
Comment (lbrynaut) [inline src/wallet/wallet.cpp]: Use static_cast

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2198,12 +2198,12 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
                     }
                     else
                     {
-                        if (nChangePosInOut == -1)
+                        if (nChangePosInOut < 0)
```
Comment (lbrynaut) [inline src/wallet/wallet.cpp]: Why change this?

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3081,7 +3081,7 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
     if (origTx.vout.size() == 0)
         throw JSONRPCError(RPC_INVALID_PARAMETER, "TX must have at least one output");
 
-    if (changePosition != -1 && (changePosition < 0 || changePosition > origTx.vout.size()))
+    if (changePosition != -1 && (changePosition < 0 || size_t(changePosition) > origTx.vout.size()))
```
Comment (lbrynaut) [inline src/wallet/rpcwallet.cpp]: Use static_cast

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (lbrynaut) [reply src/wallet/wallet.cpp]: The cast is for the unsigned comparison(s), not changing the type.

Comment (bvbfan) [reply src/wallet/wallet.cpp]: I mean static_cast<size_t>(-2) will result in insert container past end. line: 2206

Comment (bvbfan) [reply src/wallet/wallet.cpp]: Ah, forgot it. Check is mistaken.

File: src/httpserver.cpp [hunk 4]
```diff
@@ -252,7 +252,7 @@ static std::string RequestMethodString(HTTPRequest::RequestMethod m)
 /** HTTP request callback */
 static void http_request_cb(struct evhttp_request* req, void* arg)
 {
-    std::auto_ptr<HTTPRequest> hreq(new HTTPRequest(req));
+    std::unique_ptr<HTTPRequest> hreq(new HTTPRequest(req));
```
Comment (lbrynaut) [inline src/httpserver.cpp]: We're not yet using C++11.  Use boost unique ptr in the meantime.

File: src/httpserver.cpp [hunk 5]
```diff
@@ -288,7 +288,7 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
 
     // Dispatch to worker thread
     if (i != iend) {
-        std::auto_ptr<HTTPWorkItem> item(new HTTPWorkItem(hreq.release(), path, i->handler));
+        std::unique_ptr<HTTPWorkItem> item(new HTTPWorkItem(hreq.release(), path, i->handler));
```
Comment (lbrynaut) [inline src/httpserver.cpp]: Same as above.

File: src/httpserver.cpp (hunk 4 above)
Comment (bvbfan) [reply src/httpserver.cpp]: We can't use it. The checks in configure are >= 1.20, unique_ptr lands in 1.57 i'll update patch.

**Comment by lbrynaut** [comment]
@bvbfan You're right, the required boost is a bit dated.  Rather than change that, I think the slight performance hit of using boost::shared_ptr instead is the way to go instead of unique_ptr.

**Comment by bvbfan** [comment]
Furthermore i will return auto_ptr, since we can plan to switch for C++11, hopefully soon :) I compile project with C++11 and did not realize that is not C++11 :)

**Comment by lbrynaut** [comment]
@bvbfan Right, not C++11.  The claimtrie could be completely re-factored and cleaned up when we support C++11, but alas, it remains where it is for now.  It's on the roadmap (at least we've discussed it before).

**Comment by bvbfan** [comment]
Yes, i made a huge refactor of CClaimTrie not only to move db in separate class / file even CClaimCache should be refactored and renamed, as discussed before.

**Comment by lbrynaut** [comment]
@bvbfan I'm aware.  I meant a C++11 mechanical language re-factor not an organizational/architectural re-factor.

**Comment by bvbfan** [comment]
Yes, i hope after architectural refactor, step to C++11 refactor will be easiest, i hope, second time :)

**Comment by bvbfan** [comment]
You can merge this, if you does not have any other remarks. :)

**Comment by lbrynaut** [comment]
It looks good to me.  I'm waiting on any input from @kaykurokawa, otherwise may merge later this afternoon.

**Comment by kaykurokawa** [comment]
I'm a bit hesitant to merge this.  If we are modifying Bitcoin Core code like this, than there will be a lot of fixes, and we will have a bad time trying to merge from upstream which we will eventually have to do. 

I checked through some of these changes in the most recent bitcoin code base and it does indeed seem to be fixed  in the same way you did. Maybe the better way to fix it would be to cherry-pick these commits from Bitcoin Core (not saying you have to do this for this PR, but in the future, I would check to see if you aren't redoing what has already been done). 

It is important though to fix warnings for our quality of life purposes. I think we should get a process in place to say that "hey this commit can be ignored once we merge upstream", this way we can just remove ignore them and merging will be easier. Maybe something like putting in the commit comments, "NOT FOR MERGE UPSTREAM" ? We can just search and remove these commits. Or maybe these types of changes all go in to the same branch and  get merged from that branch?

**Comment by bvbfan** [comment]
I don't notice that they changed upstream, it can be cherry-pick from upstream Core for master, then ignored when merge it from my specific branch, it is OK?

**Comment by lbrynaut** [comment]
@kaykurokawa Good point.  I say we punt on this until we formally merge from upstream.  It's something that I plan to start in the next month or two, so I think it's better to not make things potentially harder later at this point.

**Comment by tzarebczan** [comment]
What's the status of this pr? @bvbfan

**Comment by lbrynaut** [comment]
@tzarebczan I'm of the opinion that it was a good exercise, but should be closed unmerged for now.  Upstream changes being pulled in are on the roadmap, which should resolve a lot of this.

**Comment by lyoshenka** [comment]
If we cherry-pick commits from upstream, I don't think we need to mark them in any special way. Git should be able to handle this, and either way Kay told me its likely the way we get up to date is by starting with the current Core code and merging in our stuff. So maybe there's no harm in merging this?

@kaykurokawa correct me if I'm wrong.

**Comment by bvbfan** [comment]
So this should be closed i think

**Pull request closed without merging**